### json 불러와서 저장하기

In [1]:
import numpy as np
import pandas as pd
import json
import jpype
import glob
from random import shuffle
from bs4 import BeautifulSoup as bs

data_path = "data"
#data_path = "data(test)"

file_list = glob.glob("%s/*.json" % data_path)

json_train=[]

shuffle(file_list)
for json_file_name in file_list:
    json_file = json.loads(open(json_file_name).read())
    json_train += json_file["articles"]

### json에서 author, author 형태소, forumid 추출

In [2]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    if cnt % 10000 == 0:
        print(cnt)
    author_pos = ["%s_%s" % (word, pos) for word, pos in mecab.pos(article["author"])]

    labeled_train.append({
            "istroll": article["is_troll"],
            "author": article["author"],
            "author_pos": author_pos,
            "author_pos_sentences" : " ".join(author_pos),
            "forumid": article["forumid"],
            "pk": article["pk"]
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train = labeled_train.set_index('pk')

labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


author                                         author_pos  \
pk                                                                        
53865           질소까스                                   [질소_NNG, 까스_NNP]   
53866        農者天下之大本                    [農者_NNG, 天下_NNG, 之_NNG, 大本_NNG]   
53868           만두만두                                   [만두_NNG, 만두_NNG]   
53869          농부고양이                                  [농부_NNG, 고양이_NNG]   
53870       닭잡고오리발끈해      [닭_NNG, 잡_VV, 고_EC, 오리발_NNG, 끈_NNG, 해_XSV+EC]   
53871         Nodoca                                        [Nodoca_SL]   
53872          나이야가라                          [나_NP, 이야_JX, 가_VV, 라_EC]   
53873          bob12                                    [bob_SL, 12_SN]   
53874         일베충살충제                      [일_NR, 베_NNG, 충살_NNG, 충제_NNG]   
53875          오래된생각                         [오래_MAG, 된_VV+ETM, 생각_NNG]   
53876   ㄹㄷㅅㅁㄹㄴㅅㅁㄹㄷㅅㅁ  [ㄹ_JKQ, ㄷ_NNG, ㅅ_NNG, ㅁ_XSN, ㄹ_JKB, ㄴ_NNG, ㅅ_N...   
53878             컴공                                     [컴_NNP, 공_NNG]   
53880         닭학의맞자오                 [닭_NNG, 학_NNG, 의_JKG, 맞_VV, 자오_EC]   
53881         일베충차우건                  [일_NR, 베_NNG, 충_NNG, 차우_VV, 건_EC]   
53882           카타리나                                         [카타리나_NNP]   
53883            김택용                                          [김택용_NNP]   
53884        epaldks                                       [epaldks_SL]   
53885           냉동홍어                                   [냉동_NNG, 홍어_NNG]   
53887         도가도비상도                       [도가_NNG, 도_JX, 비상_NNG, 도_JX]   
53888          오래된생각                         [오래_MAG, 된_VV+ETM, 생각_NNG]   
53889           다이어트                                         [다이어트_NNG]   
53890         鹿目 まどか                              [鹿_SH, 目_NNG, まどか_SL]   
53891            CIA                                           [CIA_SL]   
53893         아이고기아야                            [아이_IC, 고기_NNG, 아야_NNG]   
53894        벌레주둥이뽑기               [벌레_NNG, 주둥_NNP, 이_JKS, 뽑_VV, 기_ETN]   
53895    타이거즈Forever                             [타이거즈_NNG, Forever_SL]   
53896          노인거북이                                  [노인_NNG, 거북이_NNG]   
53897          황소트랙터                                  [황소_NNG, 트랙터_NNG]   
53898          명박x근혜                       [명박_NNG, x_SL, 근_NNG, 혜_NNG]   
53899           녹두과자                                   [녹두_NNG, 과자_NNG]   
...              ...                                                ...   
131390            OT                                            [OT_SL]   
131391        구파발갑춘이                     [구파발_NNP, 갑_NNG, 춘_NNG, 이_JKS]   
131392            치즈                                           [치즈_NNG]   
131393       닭치고酒절酒절         [닭_NNG, 치고_JX, 酒_XSN, 절_NNG, 酒_XSN, 절_NNG]   
131394            치즈                                           [치즈_NNG]   
131395            치즈                                           [치즈_NNG]   
131396         일간워스트                                  [일간_NNG, 워스트_NNP]   
131398      pankkera                                      [pankkera_SL]   
131399      3xplanet                                 [3_SN, xplanet_SL]   
131400        나이스진남친                           [나이_NNG, 스진_NNP, 남친_NNG]   
131401         민주지팡이                                  [민주_NNP, 지팡이_NNG]   
131402        그네가깐다갈                   [그네_NP, 가_JKS, 깐_VV+ETM, 다갈_NNG]   
131403          하무가면                               [하무_NNG, 가_VV, 면_EC]   
131404     깨어있는 진보시민          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131407     국가댓글원 16호             [국가_NNG, 댓글_NNG, 원_NNG, 16_SN, 호_NNBC]   
131408     깨어있는 진보시민          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131409         하프라이프                                  [하프_NNG, 라이프_NNG]   
131411     국가댓글원 16호             [국가_NNG, 댓글_NNG, 원_NNG, 16_SN, 호_NNBC]   
131412          세훈종서                                   [세훈_NNP, 종서_NNG]   
131413     깨어있는 진보시민          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131414     깨어있는 진보시민          [깨_VV, 어_

### forumid와 author을 label encoder로 숫자로 바꿔줌

In [3]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()

labeled_train["forumid"] = le.fit_transform(labeled_train["forumid"])
labeled_train["author"] = le.fit_transform(labeled_train["author"])

labeled_train

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/utils/fixes.py:64: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  if 'order' in inspect.getargspec(np.copy)[0]:


author                                         author_pos  \
pk                                                                  
53865    22925                                   [질소_NNG, 까스_NNP]   
53866     4568                    [農者_NNG, 天下_NNG, 之_NNG, 大本_NNG]   
53868    11689                                   [만두_NNG, 만두_NNG]   
53869     8413                                  [농부_NNG, 고양이_NNG]   
53870     9593      [닭_NNG, 잡_VV, 고_EC, 오리발_NNG, 끈_NNG, 해_XSV+EC]   
53871     1645                                        [Nodoca_SL]   
53872     7421                          [나_NP, 이야_JX, 가_VV, 라_EC]   
53873     2364                                    [bob_SL, 12_SN]   
53874    21132                      [일_NR, 베_NNG, 충살_NNG, 충제_NNG]   
53875    19004                         [오래_MAG, 된_VV+ETM, 생각_NNG]   
53876     4280  [ㄹ_JKQ, ㄷ_NNG, ㅅ_NNG, ㅁ_XSN, ㄹ_JKB, ㄴ_NNG, ㅅ_N...   
53878    23999                                     [컴_NNP, 공_NNG]   
53880     9715                 [닭_NNG, 학_NNG, 의_JKG, 맞_VV, 자오_EC]   
53881    21173                  [일_NR, 베_NNG, 충_NNG, 차우_VV, 건_EC]   
53882    23918                                         [카타리나_NNP]   
53883     6777                                          [김택용_NNP]   
53884     2680                                       [epaldks_SL]   
53885     7744                                   [냉동_NNG, 홍어_NNG]   
53887    10087                       [도가_NNG, 도_JX, 비상_NNG, 도_JX]   
53888    19004                         [오래_MAG, 된_VV+ETM, 생각_NNG]   
53889     9130                                         [다이어트_NNG]   
53890     4614                              [鹿_SH, 目_NNG, まどか_SL]   
53891      733                                           [CIA_SL]   
53893    17430                            [아이_IC, 고기_NNG, 아야_NNG]   
53894    13847               [벌레_NNG, 주둥_NNP, 이_JKS, 뽑_VV, 기_ETN]   
53895    24514                             [타이거즈_NNG, Forever_SL]   
53896     8117                                  [노인_NNG, 거북이_NNG]   
53897    26577                                  [황소_NNG, 트랙터_NNG]   
53898    12093                       [명박_NNG, x_SL, 근_NNG, 혜_NNG]   
53899     8155                                   [녹두_NNG, 과자_NNG]   
...        ...                                                ...   
131390    1669                                            [OT_SL]   
131391    5867                     [구파발_NNP, 갑_NNG, 춘_NNG, 이_JKS]   
131392   23668                                           [치즈_NNG]   
131393    9653         [닭_NNG, 치고_JX, 酒_XSN, 절_NNG, 酒_XSN, 절_NNG]   
131394   23668                                           [치즈_NNG]   
131395   23668                                           [치즈_NNG]   
131396   20846                                  [일간_NNG, 워스트_NNP]   
131398    3447                                      [pankkera_SL]   
131399     307                                 [3_SN, xplanet_SL]   
131400    7415                           [나이_NNG, 스진_NNP, 남친_NNG]   
131401   12858                                  [민주_NNP, 지팡이_NNG]   
131402    6150                   [그네_NP, 가_JKS, 깐_VV+ETM, 다갈_NNG]   
131403   25708                               [하무_NNG, 가_VV, 면_EC]   
131404    6892          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131407    5874             [국가_NNG, 댓글_NNG, 원_NNG, 16_SN, 호_NNBC]   
131408    6892          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131409   25790                                  [하프_NNG, 라이프_NNG]   
131411    5874             [국가_NNG, 댓글_NNG, 원_NNG, 16_SN, 호_NNBC]   
131412   15972                                   [세훈_NNP, 종서_NNG]   
131413    6892          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131414    6892          [깨_VV, 어_EC, 있_VX, 는_ETM, 진보_NNG, 시민_NNG]   
131415    2925                                [ilwar_SL, 1432_SN]   
131417     307                                 [3_SN, xplanet_SL]   
131418   24038                               [케_IC, 터_NNG, 릭_NNG]   
131419    2925                                [ilwar_SL, 1432_SN]   
131420

In [4]:
used_model = ""

### bag of words

In [5]:
from sklearn.feature_extraction.text import CountVectorizer

def make_bag_of_words(labeled_train, max_features, col_name):
    global used_model
    used_model = "bow"
    
    vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                                stop_words = None, max_features=max_features)
    
    train_data_features = vectorizer.fit_transform(labeled_train[col_name]).toarray()

    col = ["bow_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_bow = pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)
    
    labeled_train = pd.concat([labeled_train, df_bow],axis=1)
    
    return labeled_train

## LDA

In [6]:
import gensim
from gensim import corpora, models
from gensim.corpora import TextCorpus, MmCorpus, Dictionary
from nltk.corpus import stopwords
from optparse import OptionParser

def make_lda(train, keep_n, num_topics, col_name):
    global used_model
    used_model = "lda"

    data = train[col_name]

    dictionary = corpora.Dictionary(data)
    dictionary.filter_extremes(keep_n=keep_n)

    corpus = [dictionary.doc2bow(text) for text in data]

    print("Make Lda..")

    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1)
    
    num = len(train)    
    df = []
    
    for i in range(0,num):
        if i % 10000 == 0:
            print(i)
        
        temp = [i[1] for i in lda.get_document_topics(corpus[i],minimum_probability=0)]
        df.append(temp)
    
    col = ["lda_%s_%d" % (col_name, data) for data in range(0, num_topics)]
    df = pd.DataFrame(df, columns = col)
    df.index = train.index
    
    train = pd.concat([train, df], axis=1)
        
    return train

## TF-IDF

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

def make_tf_idf(train, max_features, col_name):
    global used_model
    used_model = "tf-idf"
    
    vectorizer = TfidfVectorizer(min_df = 1, max_features = max_features)
    
    train_data_features = vectorizer.fit_transform(train[col_name]).toarray()
    
    col = ["tfidf_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_tfidf = pd.DataFrame(train_data_features, columns = col, index=train.index)
    
    train = pd.concat([train, df_tfidf],axis=1)
    
    return train

In [8]:
max_features = 1000
#%time labeled_train = make_lda(labeled_train, 1000, 200, "author_pos")
#%time labeled_train = make_bag_of_words(labeled_train, max_features, "author_pos_sentences")
labeled_train = make_tf_idf(labeled_train, max_features, "author_pos_sentences")

### predictor, model 생성

In [9]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['author_pos', 'author_pos_sentences', 'author', label])

model = RandomForestClassifier(n_estimators=10, n_jobs=-1)

### cross validation

In [10]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")

cv_result = scores.mean()

print(cv_result)

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.f

0.804858083797


/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.f

## Result

 * author를 Label Encoder로 변환: 0.67331271058
 * author의 morphs를 Bag of Words(feature = 1,000 개)로 변환: 0.733333689652
 * author의 pos를 Bag of Words(feature = 1,000 개)로 변환: 0.769833271386
 * 데이터를 shuffle하고 pos를 Bag of Words(feautre = 1,000개)로 변환: 0.81172740086
 * 데이터를 shuffle하고 pos를 TF-IDF(feature = 1,000개)로 변환 : 0.804858083797
 
 
 * author의 pos를 LDA로 변환 (num_topics = 200, keep_n = 1,000) : 0.748718319834
 * 데이터를 shuffle하고 author의 pos를 LDA로 변환 (num_topics = 200, keep_n = 1,000) : 0.772313052394

In [11]:
import os
import pickle

pre = labeled_train.columns.drop(['author_pos', 'author_pos_sentences', 'author'])

if not os.path.exists(used_model):
    os.makedirs(used_model)

pickle.dump(labeled_train[pre], open("%s/author, forum_%d.p" % (used_model, max_features), "wb"), protocol = pickle.HIGHEST_PROTOCOL)